### PR TITLE
RemoteDatawaveUserService should retry requests on UnknownHostException

### PR DESCRIPTION
**Pull request by billoley**
RemoteDatawaveUserService should retry requests on UnknownHostException

See https://github.com/NationalSecurityAgency/datawave/issues/2544

### CODE REVIEW

File: web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java [hunk 1]
```diff
@@ -570,30 +575,61 @@ public void setConfig(RemoteHttpServiceConfiguration config) {
         this.config = config;
     }
 
+    /**
+     * classes that are instances of IOException that will cause DatawaveRetryHandler to retry with delay subclasses of RemoteHttpService should override this
+     * method if necessary
+     */
+    protected List<Class<? extends IOException>> getUnavailableRetryClasses() {
```
Comment (keith-ratcliffe) [inline web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java:582]: Would something like `getRetriableClasses()` be a more succinct and/or future-proof name?

File: web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java [hunk 2]
```diff
@@ -570,30 +575,61 @@ public void setConfig(RemoteHttpServiceConfiguration config) {
         this.config = config;
     }
 
+    /**
+     * classes that are instances of IOException that will cause DatawaveRetryHandler to retry with delay subclasses of RemoteHttpService should override this
```
Comment (keith-ratcliffe) [inline web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java]: Thanks to formatter, some punctuation would be helpful :-)

File: web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java [hunk 3]
```diff
@@ -570,30 +575,61 @@ public void setConfig(RemoteHttpServiceConfiguration config) {
...
+     * method if necessary
+     */
+    protected List<Class<? extends IOException>> getUnavailableRetryClasses() {
+        return Arrays.asList(ConnectException.class);
+    }
+
+    /**
+     * classes that are instances of IOException that should not cause a retry subclasses of RemoteHttpService should override this method if necessary the
```
Comment (keith-ratcliffe) [inline web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java]: Thanks to formatter, some punctuation would be helpful :-)

File: web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java (hunk 1 above)
Comment (billoley) [reply web-services/common/src/main/java/datawave/webservice/common/remote/RemoteHttpService.java:582]: All IOExceptions are retriable except for those in nonRetriableClasses

Apache HttpClient uses a delay for response codes that indicate that the service is unavailable:SC_SERVICE_UNAVAILABLE.  Datawave adds SC_NOT_FOUND

Datawave uses the same paradigm in the IOException handler to introduce a delay for IOExceptions that indicate that the service is unavailable.  Previously ConnectException.  This PR adds HostUnknownException for the call to the authorization service.

So it's not just that these are retriable.  They are classes that indicate that the service is unavailable and should be retried with a delay.

**Review by keith-ratcliffe — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]

**Pull request merged**